### ISSUE TITLE
Python bindings using PyO3

### ISSUE THREAD

**Issue by LeonarddeR** [liblouis/louis-rs#1]
Probably pretty much out of scope for now, but as soon as it is a point of discussion, consider implementing Python bindings using [PyO3](https://github.com/PyO3/pyo3).

**Comment by LeonarddeR** [comment]
I asked Claude to write a design spec. This gives a rough indication of what needs to be done.

## 1. Goal

Expose the public Rust API of the `louis` crate to Python in an idiomatic
("Pythonic") way, suitable for use by screen readers such as NVDA. Python
bindings must not perturb the existing pure-Rust crate's build, dependencies, or
consumers.

## 2. Non-goals

- Backwards-compatible C-style API matching the original `liblouis` Python
  bindings (`translateString`, integer return codes, etc.). May be added later
  as a thin Python-side shim.
- Exposing `TranslationResult.spacing` until the underlying Rust `SpacingInfo`
  struct is implemented (currently empty / `// TODO:` in `src/lib.rs`).

## 3. Repository layout

Convert the repository into a Cargo workspace. The existing crate remains the
root crate; the Python bindings live in a sibling crate `louis-py/`.

```
louis-rs/
├── Cargo.toml                # workspace + existing main crate
├── src/                      # unchanged
├── louis-py/
│   ├── Cargo.toml            # cdylib + pyo3 + path-dep on louis
│   ├── pyproject.toml        # maturin backend, project = "louis"
│   ├── src/lib.rs            # #[pymodule] fn louis
│   ├── python/louis/__init__.py  # IntFlag/Enum wrappers + re-exports
│   ├── python/louis/louis.pyi    # hand-written type stubs
│   └── tests/                # pytest suite + conftest.py
├── dictionaries/             # unchanged
└── ...
```

Root `Cargo.toml` additions:

```toml
[workspace]
members = [".", "louis-py"]
```

`louis-py/Cargo.toml`:

```toml
[package]
name = "louis-py"
version = "0.1.0"
edition = "2024"

[lib]
crate-type = ["cdylib"]

[dependencies]
louis = { path = ".." }
pyo3 = { version = "0.28", features = ["extension-module", "abi3-py38"] }
```

The main crate's `Cargo.toml` is **not** modified except for the `[workspace]`
table. `cargo build` from the repo root continues to build the existing
library and `louis` binary unchanged.

## 4. Public Python API

Module name: `louis` (PyPI distribution name: `louis`). No older PyPI release
of this name exists.

### 4.1 Module surface

```python
import louis
from louis import (
    Translator,
    Direction,
    TranslationMode,
    TranslationResult,
    LouisError,
    TableParseError,
    TranslationError,
)
```

### 4.2 `Direction`

`#[pyclass]` exposing the Rust `parser::Direction` enum. Surfaced as a Python
`enum.IntEnum`:

```python
Direction.FORWARD   # 0
Direction.BACKWARD  # 1
```

### 4.3 `TranslationMode`

Python-side `enum.IntFlag` (declared in `python/louis/__init__.py`) with one
flag per Rust `TranslationMode` variant. Bit values are stable and explicit (do
not rely on `EnumSet` internal representation):

```python
class TranslationMode(enum.IntFlag):
    NO_CONTRACTIONS    = 1 << 0
    COMPBRL_AT_CURSOR  = 1 << 1
    DOTS_IO            = 1 << 2
    COMPBRL_LEFT_CURSOR= 1 << 3
    UC_BRL             = 1 << 4
    NO_UNDEFINED       = 1 << 5
    PARTIAL_TRANS      = 1 << 6
```

The Rust shim accepts a `u32` of OR-ed bit values and converts it into the Rust
`TranslationModes` (which is `EnumSet<TranslationMode>`).

### 4.4 `Typeform`

Plain `dict[str, str]`. The `typeforms` argument is `list[dict[str, str]] |
None` and maps directly to `Vec<HashMap<String, String>>` on the Rust side.

### 4.5 `Translator`

`#[pyclass(frozen)]` wrapping `louis::Translator`. Immutable after
construction; safe to share across threads.

```python
Translator(
    tables: list[str | os.PathLike],
    direction: Direction = Direction.FORWARD,
)

t.translate(text: str) -> str

t.translate_with_options(
    text: str,
    *,
    mode: TranslationMode = TranslationMode(0),
    typeforms: list[dict[str, str]] | None = None,
    cursor_pos: int | None = None,
) -> TranslationResult
```

- `tables` accepts any `os.PathLike` or `str`; each is converted to `PathBuf`.
- Table resolution honors `LOUIS_TABLEPATH` via the existing `search_path`
  crate behavior — no extra Python-side logic.
- The constructor calls `Translator::new` inside `py.allow_threads(...)`;
  table parsing can be slow.
- Both `translate` methods release the GIL via `py.allow_threads(...)`.

Method naming mirrors the Rust API 1:1 (`translate` / `translate_with_options`)
rather than collapsing into a single overloaded method, for clarity and easier
documentation mapping.

### 4.6 `TranslationResult`

`#[pyclass(frozen, get_all)]` with read-only attributes mirroring the Rust
struct:

```python
result.output: str
result.typeforms: list[dict[str, str]] | None
result.output_positions: list[int] | None
result.input_positions: list[int] | None
result.cursor_pos: int | None
```

A `__repr__` is provided for debugging. `spacing` is **not** exposed yet
(Rust-side `SpacingInfo` is empty).

### 4.7 Exceptions

Created via `pyo3::create_exception!`:

```
LouisError(Exception)
├── TableParseError(LouisError)   # has .errors: list[str]
└── TranslationError(LouisError)
```

`From<louis::TranslationError> for PyErr` dispatches:

- `TranslationError::ParseFailed(errs)` → `TableParseError`, with `errors`
  attribute set to `[format!("{e}") for e in errs]`.
- `TranslationError::TranslationFailed(e)` → `TranslationError` with the
  formatted message.

### 4.8 Type stubs

`python/louis/louis.pyi` is hand-written and shipped in the wheel. The public
surface is small enough that manual maintenance is cheaper than codegen.

## 5. Data flow

```
Python call
  → Rust shim converts arguments (list[PathLike] → Vec<PathBuf>,
                                   IntFlag bits  → TranslationModes,
                                   list[dict]    → Vec<HashMap<String,String>>)
  → py.allow_threads(|| louis::Translator::translate_with_options(...))
  → result struct converted back to PyObject (TranslationResult)
Errors flow back through From<louis::TranslationError> for PyErr.
```

## 6. Threading model

`Translator` is constructed once and is immutable (`#[pyclass(frozen)]`). The
Rust struct's interior is read-only after `Translator::new`; no synchronization
is needed for concurrent `translate` calls. GIL release is therefore safe and
correct, and is preferable for multi-threaded use.

## 7. Testing

`louis-py/tests/`:

- `conftest.py` — sets `LOUIS_TABLEPATH` to the repo `dictionaries/` directory.
- `test_translate.py` — basic forward translate with a known table; verify
  output string.
- `test_options.py` — `translate_with_options` populates `TranslationResult`;
  `cursor_pos` echoed; `output_positions` length matches input char count.
- `test_modes.py` — `TranslationMode` flags compose with `|`, are passed
  through and accepted by the shim.
- `test_errors.py` — bogus table path raises `TableParseError`; assert
  `isinstance(exc, LouisError)`; assert `exc.errors` is a non-empty list.
- `test_threading.py` — N concurrent threads call `translate`; assert no
  deadlock and identical outputs (smoke test for GIL release correctness).

Local workflow: `cd louis-py && maturin develop && pytest`.

## 8. Build, CI, and distribution

### 8.1 Local build

```
cd louis-py
maturin build --release          # produces wheels in target/wheels/
maturin develop                  # editable install for tests
```

### 8.2 CI

New workflow `.github/workflows/python-wheels.yml` using
`PyO3/maturin-action`:

- **Trigger on PR** when `louis-py/**` changes: build wheel for
  `ubuntu-latest`, run `pytest`. No publish.
- **Trigger on tag `py-v*`**: full matrix build:
  - `windows-latest` × `x86_64`, `x86` (32-bit, required for current NVDA)
  - `ubuntu-latest` × `x86_64` (manylinux2014), `aarch64`
  - `macos-latest` × `x86_64`, `aarch64`
  Plus `sdist`. Publish to PyPI via `pypa/gh-action-pypi-publish` configured
  with a Trusted Publisher.

`abi3-py38` means one wheel per (OS, arch) covers Python 3.8+.

### 8.3 Versioning

`louis-py/pyproject.toml` carries an independent version (Python bindings may
iterate faster than the Rust crate). Release tag scheme: `py-v<semver>`, e.g.
`py-v0.1.0`. The Rust crate keeps its own tag scheme.

## 9. Out of scope (deferred)

- Exposing `TranslationResult.spacing` — Rust `SpacingInfo` struct is empty.
  Expose when populated upstream.
- C liblouis Python-API compatibility shim (`translateString`, return codes).
  Separable, can be added later as a pure-Python module sitting on top of
  these bindings.

## 10. Open risks

- `LOUIS_TABLEPATH` on Windows: confirm `search_path` honors Windows-style
  semicolon-separated paths. Validate in `test_translate.py`.
- `enumset` bit values: do not assume they match the explicit `1 << n`
  values we expose. The Rust shim must perform an explicit
  `TranslationModes::insert(...)` per bit set in the Python `u32`, not a raw
  `from_bits_truncate`.
- `louis::Translator` `Send + Sync`: not asserted explicitly in
  `src/translator.rs`. Auto-derived from field types. Verify at implementation
  time with a `static_assertions::assert_impl_all!` test; if it does not hold,
  wrap in `Mutex` (cheap — calls are short, contention low) and document the
  change.
- Pinned `pyo3 = "0.28"` is indicative. Implementer should pick the latest
  stable pyo3 at the time of implementation; API changes between 0.22 and
  newer versions are minor and limited to attribute syntax.

**Comment by LeonarddeR** [comment]
I have also considered [UniFFI](https://github.com/mozilla/uniffi-rs). For Python, it creates ctypes bindings. As Python use for liblouis is a major use case (NVDA, Orca, etc.), I still believe that we should go the PyO3 route.

**Comment by LeonarddeR** [comment]
@egli Would you like me to work on this? I got some hours and AI credits to spare this summer. Really like to see this project going forward, since I think in the end it is a good candidate to replace liblouis in NVDA.

* Re point 3 in design spec, Repo layout is also related to https://github.com/liblouis/louis-rs/pull/7 and other future additions that might need new crates.
* Re point 4. To me louis feels as the most natural name for the python package, but that conflicts with the current ctypes based bindings for liblouis. We could also consider pylouis, louispy or similar.

**Comment by egli** [comment]
> [@egli](https://github.com/egli) Would you like me to work on this?

Hi @LeonarddeR , yes, I'd like you to work on this. I think I cleaned up the API somewhat (have a look at lib.rs), so it should be more or less ready. In particular I replaced the `typeform` with an `Vec<EmphasisSpan>`

But generally I think working on bindings should also uncover some gaps or problems in the API. It is not set in stone. If you have suggestions go ahead.

For example I notice right now that the `TranslationOptions` struct has pub on all fields. Usually I don't like that. Might tighten that.

**Comment by egli** [comment]
Follow-up on the `TranslationOptions` note above: tightened it in e307a57 — fields are now private, constructed via `TranslationOptions::default().with_mode(...).with_emphasis(...).with_cursor_pos(...)` and read via `mode()`/`emphasis()`/`cursor_pos()` accessors. Worth keeping in mind for the Python shim's `translate_with_options` argument conversion.

**Comment by LeonarddeR** [comment]
This is live in https://github.com/LeonarddeR/louis-py
@egli Happy to transfer it over if you like.

**Comment by egli** [comment]
> This is live in https://github.com/LeonarddeR/louis-py [@egli](https://github.com/egli) Happy to transfer it over if you like.

Not sure how this transfer works. Sure I'd like that, can you do it?

**Comment by LeonarddeR** [comment]
I can transfer to the liblouis organization and I am somehow a member but I have insufficient privileges. I think I temporarily need the create repo privilege for that.

**Comment by egli** [comment]
I enabled it temporarily

**Comment by LeonarddeR** [comment]
@egli Thanks, it went through. See https://github.com/liblouis/louis-py
